### ISSUE TITLE
Did not match ack and chunk

### ISSUE THREAD

**Issue by khachaturh** [tkrs/poston#111]
Once Failed to read response, many same errors and twice write occur after that. Here are logs:

```
2020-09-15T13:25:54.035897792+00:00 - [fluent-worker-pool][WARN] - Failed to read response, chunk: OTk3ZDcxNTktNjUzMC00NTZmLTljN2UtODliMzhkODE1NmUx, cause: Transient(NetworkError("Resource temporarily unavailable (os error 11)"))
2020-09-15T13:26:04.591909981+00:00 - [fluent-worker-pool][WARN] - Failed to read response, chunk: OTk3ZDcxNTktNjUzMC00NTZmLTljN2UtODliMzhkODE1NmUx, cause: Transient(NetworkError("Resource temporarily unavailable (os error 11)"))
2020-09-15T13:26:06.540205861+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: OTk3ZDcxNTktNjUzMC00NTZmLTljN2UtODliMzhkODE1NmUx, chunk: YzAyZTVhYjgtYjRmNS00Y2JmLWEwMzYtMGQ1YzRkZGU1NWNl
2020-09-15T13:26:09.714639498+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: NDNhNjFhY2MtZTlhMC00NWQ0LThkNGYtOWVmZTkyY2UyM2Q0, chunk: ZGEwNzMwZjktZTVmMC00ZDE5LWFmMmMtMzY1NWJkZTZhOGEx
2020-09-15T13:26:13.858623537+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: ODAzYzA4ODQtMjdjYS00NTU2LTk2ZTgtZGZiYTE0OGU1YWIz, chunk: ZWI1YmU4NzItYTNlNS00ZDE5LTlkMTQtMjUxMGY3MWQ0MGIw
2020-09-15T13:26:14.173476908+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: ZTBhODZmYjItOGE3NC00MjA0LTkwMzAtYzBkZTY5YTExMDYz, chunk: MDhmNzI1NWUtNmY0YS00MzU1LTk3ZDUtOTBmY2NmNDI4MTBi
2020-09-15T13:26:14.298042783+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: MDhmNzI1NWUtNmY0YS00MzU1LTk3ZDUtOTBmY2NmNDI4MTBi, chunk: MjU0NTA1MTctMDA4Ny00NjRjLTkzM2QtNGUzOWUzNzBkMjU4
2020-09-15T13:26:14.438972065+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: MjU0NTA1MTctMDA4Ny00NjRjLTkzM2QtNGUzOWUzNzBkMjU4, chunk: YmRjZjM3OGUtOWJkYi00NWI5LWJlMjYtNmNiNDI0MDYwZmEw
2020-09-15T13:26:14.656711886+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: YmRjZjM3OGUtOWJkYi00NWI5LWJlMjYtNmNiNDI0MDYwZmEw, chunk: OGQzNDkyNjMtNmQ2Yy00OTc0LWIzZjItZTg3NzU3OTJlNDUx
2020-09-15T13:26:14.810611582+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: OGQzNDkyNjMtNmQ2Yy00OTc0LWIzZjItZTg3NzU3OTJlNDUx, chunk: ZjYxMjIyMjctZTUyOC00MjUzLThjZmEtNGZlNDRlYzc2MGQy
2020-09-15T13:26:14.894403624+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: ZjYxMjIyMjctZTUyOC00MjUzLThjZmEtNGZlNDRlYzc2MGQy, chunk: MzNhYWMzMDktMGU3OC00Y2Q3LThkOGEtNGM0MmVlNDZmZDM0
2020-09-15T13:26:15.301887782+00:00 - [fluent-worker-pool][WARN] - Did not match ack and chunk, ack: NzVkY2M1ZTYtMDMyNC00NTk0LThmNDItYjliMmJlYmJiNzRh, chunk: MzdjZTM5ZmYtOWZmNi00MTZmLTkxY2YtMWE3ODQxNWVhMmYz
```

There are 2 bugs. 

1. The error "Resource temporarily unavailable (os error 11)" should be handled differently, or should be ignored. It means we will receive the response, but with a delay.
1. After this case all messages are sent several times. It depends on the settings `write_retry_timeout` and `read_retry_timeout`
by default `write_retry_timeout` is 30 secs, `read_retry_timeout` 10 secs. Messages are sent twice for these settings. 
I see nested closure with backoff pattern in the method
`fn write_and_read(&mut self, buf: &[u8], chunk: &str) -> Result<(), Error>`, file https://github.com/tkrs/poston/blob/master/src/connect.rs. 
External closure with write works successfully once, and nested read_backoff is failed after 10 secs. It means that external backoff also is failed, but writes to socket. So it will write again the same message on next backoff step because it has a 30 secs timeout. I thinк `write_retry_timeout` should be less or equal than `read_retry_timeout`. 
Anyway the nested backoff is hard to understand and the kind of anti pattern. May be needs refactoring `write_and_read` function.

Thanks

**Comment by tkrs** [comment]
Thanks for reporting. Is the version 0.7.0?

**Comment by khachaturh** [comment]
> Thanks for reporting. Is the version 0.7.0?

Yes, version is 0.7.0

**Comment by tkrs** [comment]
Hmm, wasn't any error/warn logs output before the above logs?

**Comment by khachaturh** [comment]
> Hmm, wasn't any error/warn logs output before the above logs?

No, the first log is "Failed to read response"

**Comment by tkrs** [comment]
Sorry for the late reply. I have published an attempt to solve this as 0.7.1.
